### ISSUE TITLE
New version to support quarkus >= 3.3.x

### ISSUE THREAD

**Issue by DanielScLN** [quarkiverse/quarkus-file-vault#98]
Hi everyone,

first of all. I like this extension very much. I also want to update to the current release of quarkus 3.9.3 but the depracted enum ConfigPhase.BOOTSTRAP is still on your release version 2.0.0. I see you already fixed that in your main branch. I guess only a new version should be published, right? Would be very great to hear from you.

regards

**Comment by sberyozkin** [comment]
@DanielScLN Sorry, I've missed this ping, I can try to release soon. Note though, we recommend use the Quarkus configuration system directly: https://smallrye.io/smallrye-config/Main/config/secret-keys/#secret-keys-expressions

CC @radcortez

**Comment by radcortez** [comment]
Correct. Most of the functionality provided by this extension is now available in SmallRyeConfig directly:

https://smallrye.io/smallrye-config/3.9.1/config/secret-keys/
https://smallrye.io/smallrye-config/3.9.1/config-sources/keystore/

We would love to hear some feedback.

**Comment by sberyozkin** [comment]
Fixed by #110, the actual release is pending, depending on what we decide about supporting this extension. Perhaps it can be kept open to the community contributions only if anyone is already using it.

**Comment by claudioweiler** [comment]
Hi @radcortez, @sberyozkin 

I take a look ate smallrye.config.source, but I didn't found the CredentialProvider implementation. Or I'm missing something? Should I make my own custom provider?

Your extension is really good at handling credential provider for databases. Can you consider doing a new release?

Thanks!

**Comment by radcortez** [comment]
You can use https://smallrye.io/smallrye-config/Latest/config-sources/keystore/ combined with a https://smallrye.io/smallrye-config/Latest/config/secret-keys/. You shouldn't need a `CredentialProvider`. Just encrypt and store your secrets in the keystore, and use the secret handler expression to reference these. Also check https://quarkus.io/guides/config-secrets

**Comment by claudioweiler** [comment]
Hi @radcortez , thanks for your return, but... Sorry...

Read again all documentations, but I can't find it, nor examples at web.

What I understand about secret handler: a handler to decrypt secrets.

My keystore config is ok:

```properties
smallrye.config.source.keystore.db.path = file.p12
smallrye.config.source.keystore.db.password = dbvaultsecret
```

But to get secrets, I need to use Config:

```java
@Inject
Config config;
... 
config.getValue("aliasinvault", String.class);
```

But what I need is to inject password at database configuration (quarkus.datasource...) like this extension does.

So far, could only inject those values with a custom provider that implements above java code:

```properties
# custom credential provider created to map "je.vault"
quarkus.datasource.db.credentials-provider = je.vault.provider.db
# retrieving password with an alias
je.vault.provider.db.alias = aliasinvault
```

Thanks again!

**Comment by radcortez** [comment]
In this case you don't need the credentials-provider. You use:

```properties
quarkus.datasource.password=${secret-handler::value}
```

**Comment by claudioweiler** [comment]
Thanks again, @radcortez , but I lacking something. I believe that, in your example, "value" should be my alias, but have no clue about what is "secret-handler".

Anyway, it's out of scope of this thread to ask for support, so I will try to make a question on stackoverflow.

Thanks!

**Comment by radcortez** [comment]
Do this:

Add `smallrye-config-crypto` and `smallrye-config-source-keystore` to the dependencies.

Generate an encrypted value of the database password:

`quarkus config set --encrypt quarkus.datasource.password password`

This will automatically add to `application.properties`:

```properties
quarkus.datasource.username=database
quarkus.datasource.password=${aes-gcm-nopadding::DIBctwqzU58DG3IJJ8RSILIW_koEUb3jEN-G1L0fXu7Bii2x8Q}

smallrye.config.secret-handler.aes-gcm-nopadding.encryption-key=ZALNK91dzZy9ez1WVCMQ7A
```

Now, move the value to a keystone:

```shell
echo DIBctwqzU58DG3IJJ8RSILIW_koEUb3jEN-G1L0fXu7Bii2x8Q | keytool -importpass -alias quarkus.datasource.password -keystore properties -storepass arealpassword -storetype PKCS12 -v
```

You can now remove the `quarkus.datasource.password` and use the keystore:


```properties
quarkus.datasource.username=database

smallrye.config.source.keystore."properties".path=properties
smallrye.config.source.keystore."properties".password=arealpassword
smallrye.config.source.keystore."properties".handler=aes-gcm-nopadding

smallrye.config.secret-handler.aes-gcm-nopadding.encryption-key=ZALNK91dzZy9ez1WVCMQ7A
```

Quarkus reads the keystore, and since `quarkus.datasource.password` is inside, it will automatically decrypt it and use it.

Remember, you still need to protect the keystore password, plus the encryption key:
https://quarkus.io/guides/config-secrets#protect-the-keystore-password

**Comment by claudioweiler** [comment]
Oh, I see we have a misunderstanding here, @radcortez .

I want to still use my PKCS12 file as keystore: `smallrye.config.source.keystore.db.path = file.p12` (or at least, some other way)

This works with this extension, and with my custom CredentialProvider. This approach solves my problem with securing both "keystore" and "keystore password" (they are not part of commits).

Maybe I can also use Secret Keys to do more security to keystore password, but what I want most is to use keystore as CredentialProvider to database: `quarkus.datasource.db.credentials-provider`.

Thanks!

**Comment by claudioweiler** [comment]
Just to feedback my shame.

It's SO simple after I analyse it calmly.

Sorry, @radcortez , it's all in your last post, but I didn't see it.

Just need to reference alias (inside keystore) in database password:

```properties
smallrye.config.source.keystore.secrets.path = ${VAULT_PATH}
smallrye.config.source.keystore.secrets.password = ${VAULT_SECRET}

quarkus.datasource.db1.password = ${${VAULT_DB1_ALIAS}}
```

Thanks your patience!

**Comment by radcortez** [comment]
Yes, that is an added benefit of how the SR Config Keystore works. We call it a Configurable ConfigSource, because it requires configuration to bootstrap, but once that is done, it contributes configuration like any other source. In this case, it is more useful, because it can be used to add any configuration and does not require any additional setup to read.

I'm glad it worked out for you :)
